aptos-move/framework/src/natives/cryptography/groups/arkworks_backend.rs

```diff
@@ -0,0 +1,1896 @@
+    let g1_handles = pop_arg!(args, Vec<u64>);
+    match (
+        type_tag_0.as_str(),
+        type_tag_1.as_str(),
+        type_tag_2.as_str(),
+    ) {
+        ("0x1::groups::BLS12_381_G1", "0x1::groups::BLS12_381_G2", "0x1::groups::BLS12_381_Gt") => {
+            let g1_prepared: Vec<ark_ec::models::bls12::g1::G1Prepared<Parameters>> = g1_handles
+                .iter()
+                .map(|&handle| {
+                    let element = context
+                        .extensions()
+                        .get::<Bls12381Context>()
+                        .get_g1_point(handle as usize);
+                    ark_ec::prepare_g1::<ark_bls12_381::Bls12_381>(element.into_affine())
+                })
+                .collect();
+            let g2_prepared: Vec<ark_ec::models::bls12::g2::G2Prepared<Parameters>> = g2_handles
+                .iter()
+                .map(|&handle| {
+                    let element = context
+                        .extensions()
+                        .get::<Bls12381Context>()
+                        .get_g2_point(handle as usize);
+                    ark_ec::prepare_g2::<ark_bls12_381::Bls12_381>(element.into_affine())
+                })
+                .collect();
+
+            let z: Vec<(
+                ark_ec::models::bls12::g1::G1Prepared<Parameters>,
+                ark_ec::models::bls12::g2::G2Prepared<Parameters>,
+            )> = g1_prepared
+                .into_iter()
+                .zip(g2_prepared.into_iter())
+                .collect();
+            let result = ark_bls12_381::Bls12_381::product_of_pairings(z.as_slice());
+            let result_handle = context
+                .extensions_mut()
+                .get_mut::<Bls12381Context>()
+                .add_gt_point(result);
+            Ok(NativeResult::ok(
+                (gas_params.bls12_381.g1_affine_to_prepared + gas_params.bls12_381.g2_affine_to_prepared + gas_params.bls12_381.pairing_product_per_pair) * NumArgs::new(g1_handles.len() as u64) + gas_params.bls12_381.pairing_product_base,
+                smallvec![Value::u64(result_handle as u64)],
+            ))
+        }
+        _ => {
+            Ok(NativeResult::err(InternalGas::zero(), abort_codes::E_UNKNOWN_PAIRING))
+        }
+    }
+}
+
+fn hash_to_bls12381_fr(bytes: &[u8]) -> Fr {
+    let mut digest = Sha256::digest(bytes).to_vec();
+    digest[31] = 0;
+    Fr::from_random_bytes(digest.as_slice()).unwrap()
+}
+
+fn hash_to_element_internal(
+    _gas_params: &GasParameters,
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert_eq!(1, ty_args.len());
+    let type_tag = context
+        .type_to_type_tag(ty_args.get(0).unwrap())?
+        .to_string();
+    let bytes = pop_arg!(args, Vec<u8>);
+    match type_tag.as_str() {
+        "0x1::groups::BLS12_381_G1" => {
+            //TODO: replace this naive implementation.
```

Comment:
Yep, we cannot use this naive implementation because one can easily compute discrete logs w.r.t. the generator for a hashed element.

This makes certain applications impossible, like BLS signature verification.

Reply:
If there is no implementation available in `arkworks`, we can use `blst`'s maybe.

Or, we can abort.

You will need a new abort code for such functions that are not implementable yet.